File: src/main/java/org/customer/service/TransactionService.java
package org.customer.service;


import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.Getter;
import org.customer.cache.RedisConfig;
import org.customer.dto.TransactionDTO;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import org.springframework.context.annotation.Bean;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.kafka.annotation.KafkaListener;

import org.springframework.stereotype.Service;
import java.util.*;


@Getter
@Service
public class TransactionService extends CustomerService {

  public TransactionService(RedisConfig redisTemplate) {
    super(redisTemplate);
  }

  @KafkaListener(topics = "transactions", groupId = "transaction_group")
  public void consumeTransaction(String transactionJson) {
    TransactionDTO transaction = null ;
    try { //handle transactions
      ObjectMapper objectMapper = new ObjectMapper();
      transaction = objectMapper.readValue(transactionJson, TransactionDTO.class);

      this.redisTemplate.transactionRedisTemplate().opsForList().rightPush("TRANSACTIONS:" + transaction.getAccountId(), transaction);
    } catch (Exception e) {
      logger.error("Transaction record could not consumed -> {} with error : {}", transaction, e.getMessage());
    }
  }

  public List<TransactionDTO> getTransactionsByAccount(String accountId) {
    return redisTemplate.transactionRedisTemplate().opsForList().range("TRANSACTIONS:" + accountId, 0, -1);
  }
}

File: src/main/java.kotlin/org/customer/service/TransactionService.kt
package org.customer.service

import com.fasterxml.jackson.databind.ObjectMapper
import lombok.Getter
import org.customer.cache.RedisConfig
import org.customer.dto.TransactionDTO
import org.springframework.kafka.annotation.KafkaListener
import org.springframework.stereotype.Service


@Getter
@Service
class TransactionService(redisTemplate: RedisConfig?) : CustomerService(redisTemplate) {
    @KafkaListener(topics = ["transactions"], groupId = "transaction_group")
    fun consumeTransaction(transactionJson: String?) {
        var transaction: TransactionDTO? = null
        try { //handle transactions
            val objectMapper = ObjectMapper()
            transaction = objectMapper.readValue(transactionJson, TransactionDTO::class.java)

            redisTemplate.transactionRedisTemplate().opsForList()
                .rightPush("TRANSACTIONS:" + transaction.getAccountId(), transaction)
        } catch (e: Exception) {
            logger.error("Transaction record could not consumed -> {} with error : {}", transaction, e.message)
        }
    }

    fun getTransactionsByAccount(accountId: String?): MutableList<TransactionDTO?>? {
        return redisTemplate.transactionRedisTemplate().opsForList().range("TRANSACTIONS:$accountId", 0, -1)
    }
}